### DISCUSSION TITLE
Setting a particular tab to focus when the docs are built.

### DISCUSSION THREAD

**Discussion by UltraInstinct05** [squidfunk/mkdocs-material#3982] [Q&A] [answered]
Is there a way to focus on a given tab ?

As an example, take https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#grouping-code-blocks as a reference. There are 2 tabs, `C` & `C++`. When the docs are built, I'd like to focus on `C++` instead of `C` i.e. ability to focus a different tab other than first.

**Comment by squidfunk** [comment, marked as answer]
@facelessuser I thought this was possible with something like `===+` or `===!`, similar to how details work, but can't seem to find something on your docs!? If not, would this be something to consider? Also couldn't find a reference in the [tests](https://github.com/facelessuser/pymdown-extensions/blob/main/tests/test_extensions/test_tabbed.py).

**Reply by facelessuser** [reply]
`===!` is used to force a new tab set if you have them directly adjacent:

````
=== "Tab 1"
    Markdown **content**.

    Multiple paragraphs.

=== "Tab 2"
    More Markdown **content**.

    - list item a
    - list item b

===! "Tab A"
    Different tab set.

=== "Tab B"
    ```
    More content.
    ```
````

There has never been a way to force a tab selection.  We could consider an addition.

With all of that said, if the user in the OP wants to focus on C++ instead of C, why not specify C++ first? We can add some kind of tab selector with `===+`, but generally, at least to me, it would make more sense to just specify what you intend as the main focus first 🤷🏻.

**Reply by squidfunk** [reply]
Valid concerns. However, it would probably be beneficial to at least have the option to preselect a tab. As I understand, it would only influence which radio button receives the `checked` attribute. In any case, this is something, which should be discussed over at [Python Markdown Extensions](https://github.com/facelessuser/pymdown-extensions), so @Ultra-Instinct-05 if you feel this is something you need, please take this upstream 😊

**Reply by UltraInstinct05** [reply]
My main use case is that I have a bunch of tabs in a page that are logically arranged e.g. `Tab 1`, `Tab 2`, `Tab 3` etc.

As I progress further into the page explaining various concepts & building up content material present in the tabs, I'd like to focus on a specific tab e.g. In a given section, explaining content in `Tab 2` is relevant, so it should get the focus, but I don't want to rearrange Tabs just so that `Tab 2` is first, otherwise, that would look weird.

**Reply by facelessuser** [reply]
Like I said, I have no problem considering such an option regardless. Without context, my advice is to order things in importance, but now that context has been provided, I can see that in this case specifying a default tab may make sense.

**Reply by UltraInstinct05** [reply]
Opened an issue to track it https://github.com/facelessuser/pymdown-extensions/issues/1724

**Reply by squidfunk** [reply]
Also possibly related:

- #3976

**Reply by UltraInstinct05** [reply]
Thanks for reminding. I noticed I never gave a use case when I opened the discussion, so have done it now https://github.com/squidfunk/mkdocs-material/issues/3976#issuecomment-1146546243

**Reply by facelessuser** [reply]
While I'm looking to roll this out this weekend, Material will break this behavior when `content.tabs.link` is enabled. I'm trying to figure out how I can work around this to demonstrate it in the docs as I currently have `content.tabs.link` enabled 😞. I may just have to disable it and not use the linked content tabs feature...

**Reply by UltraInstinct05** [reply]
I guess what the `mkdocs-material` page https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#linked-content-tabs could be adapted i.e. demonstrating using PNG's ?

**Reply by squidfunk** [reply]
Happy to fix any issues with `content.tabs.link`, if you can provide a reproduction showing the error.

**Reply by facelessuser** [reply]
I'm not sure it's a bug as much as the features are incompatible. I have a branch up. I thought I'd get it in the next release, but I may hold off.

**Reply by facelessuser** [reply]
Here's the branch for anyone wanting to try it: https://github.com/facelessuser/pymdown-extensions/pull/1725.  Syntax is simply:

```
===+ "Selected tab"
    content
```

`content.tabs.link` force linked tabs to be selected instead of the specified tab with `checked`.

**Reply by squidfunk** [reply]
Okay, so I had some time to think about this and think you're right, the features are incompatible in a way that if one of the content tabs that is part of a `tabbed` container was previously selected by the user, Material for MkDocs will always select that tab. The idea is that linking of content tabs only really makes sense if you do it across pages, e.g. when you have API documentation with different languages. It's also the reason why we don't have it enabled on our documentation, as it's not a good fit for mixed documentation.

We could allow to enable linking of content tabs selectively (as we allow it for [code annotations](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#code-annotations)) if only we could add attributes or classes to content tabs. This is currently not supported, or I haven't found a way how it works.

<img width="711" alt="Bildschirmfoto 2022-06-05 um 08 20 51" src="https://user-images.githubusercontent.com/932156/172038120-8e36837f-b93d-41a5-aab8-20242432b7fb.png">

Maybe to improve the understanding of the current implementation of linked content tabs, here's how it works: if the user clicks a tab, we extract the text content from that tab (e.g. ignoring icons or other non-textual elements) and push that to the front of a list which we store in local storage. This is done each time the user clicks a tab. Now, when a content tab container is mounted, the list is checked for the first occurrence which can be matched against one of the tabs, and then that tab is selected. Think of it as "always select the tab that the user clicked last".:

https://github.com/squidfunk/mkdocs-material/blob/30dcb66c710b960d048ae6de5cfdd97782a714fd/src/partials/javascripts/content.html#L24-L39

Persistence is handled here:

https://github.com/squidfunk/mkdocs-material/blob/30dcb66c710b960d048ae6de5cfdd97782a714fd/src/assets/javascripts/components/content/tabs/index.ts#L191-L209

**Reply by facelessuser** [reply]
Yeah, the easiest way to allow classes would be to use `attr_list` via `=== "title" {.class}`. Admonitions and details are highly dependent on classes, so classes are built into their syntax via `!!! class "title"`, but it makes less sense for tabs.

Unfortunately, when we created both Details and Tabbed, we based it on the naive regex of Admonitions which assumes that anything between the first quote and the quote at the end of the line is considered the title. This is sloppy and allows the following:

```
!!! note "I can add "whatever" I want"
```

On the surface, this is fine, but it makes the regex difficult to extend to the example below without breaking legacy expectations. You then need to parse the title properly and probably require quotes inside a string to be escaped, this is the part that may catch people unawares:

```
!!! note "I can add \"whatever\" I want" {.class}
```

This is probably what _should_ be done, but it's definitely an approach that would break stuff.

I guess you could do something like code blocks do and match the start and end quotes `""I can add "whatever" I want""`. This idea was rejected when I proposed it for Admonitions though. If I were to come up with an approach, I'd like it to be some approach that could be unified across all the like Admonition approaches.

Anyways, I could add the feature regardless, it doesn't require Material to support both tab linking and defaulted tabs, I would just need to think about how to demonstrate the idea in the docs. Maybe I would just use Codepen 🤷🏻.
